docs/contributing/999-rfcs/2024-02-15-resource-pinning.md

```diff
@@ -0,0 +1,249 @@
+- The pinfile is stable and predictable between compilations on different machines
+- CLI provides clear error messages and guidance for resolving deletion/moves of pinned resources
+
+### Cases
+
+For the following cases, assume a user has expressed their intent to protect a resource from deletion/moves somehow.
+The following cases should be handled by this RFC:
+
+- A pinned resource's id was changed in source (using `as`)
+- A pinned resource was moved to a different path in the resource tree (e.g. placed inside a different class)
+- An update to wing itself changed a pinned resource's path
+- An update to a dependency changed the id's within a resource it exports, and the user has pinned an instance of that resource
+
+## User Journey
+
+Sally Stateful creates a wing file with a bucket
+
+```wing
+// main.w
+bring cloud;
+
+new cloud.Bucket() as "CoolBucket";
+```
+
+They run `wing compile` and then `wing test`. Nice, looks like everything works.
+
+Running `wing compile -t tf-aws` also succeeds with the following output and a new file called main.w.pin.json appears:
+
+```shell
+# (output)
+[+pin] aws_s3_bucket.CoolBucket
+```
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "aws_s3_bucket.CoolBucket": {
+        "type": "@cdktf/provider-aws.s3bucket.S3Bucket",
+        "attributes": {
+          "bucket_prefix": "coolbucket-c87e9a4d-",
+          "force_destroy": false
+        }
+      }
+    }
+  }
+}
+```
+
+Sally deploys this to AWS and everything is great.
+
+Some time later, Sally decided to move the bucket into a new abstraction.
+
+```wing
+// main.w
+bring cloud;
+
+class MyBucket {
+  bucket: cloud.Bucket;
+
+  new() {
+    this.bucket = new cloud.Bucket();
+  }
+}
+
+let myBucket = new MyBucket();
+```
+
+`wing compile` and `wing test` still work as expected with simulator output.
+
+`wing compile -t tf-aws` however fails with an error message
+
+```shell
+# (output)
+[-pin] aws_s3_bucket.CoolBucket
+[+pin] aws_s3_bucket.MyBucket_AD8CE4AC
+
+Error: The following pinned resource has been moved or deleted:
+  - aws_s3_bucket.CoolBucket
+
+If moved, update the key in the pinfile to reflect the new path.
+If deleted, remove the entry from the pinfile.
+
+```
+> In the future we will add support for `wing pin mv` and `wing pin rm`
+
+This refactor changed the location of the bucket in the generated terraform, but more importantly the name/prefix of the bucket to be deployed.
+This change is a destructive operation that required replacement.
+
+For now, Sally wants to avoid this destruction so she updates the pinfile to reflect the new path while adding an "originalPath" for the original.
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "aws_s3_bucket.MyBucket_AD8CE4AC": {
+        "type": "@cdktf/provider-aws.s3bucket.S3Bucket",
+        "originalPath": "aws_s3_bucket.CoolBucket",
```

Comment:
This still feels wrong... I am sorry :-(

The new terraform resource name (`aws_s3_bucket.MyBucket_AD8CE4AC`) is not really something that I care about or understand as a user and not something that I need to carry forward with me.

Wing code "speaks" in logical node path terms. These are the IDs that users understand (and oftentimes assign themselves through `as X`).

I am sorry but I still think that this simply needs to be a map between *logical node paths* and *terraform resource names*.

Happy to talk more about this because it feels like something is misaligned in requirements/use case :-)

I also don't think we need the attributes of the resource here. This will be huge duplication with the Terraform output and I don't see why it is needed.

This is what I think this should look like:

```json
// main.w.pin.json
{
  "version": "1",
  "pinned": {
    "tf-aws": {
      "/MyBucket/Bucket": {
        "terraform_name": "aws_s3_bucket.CoolBucket",
        "type": "@cdktf/provider-aws.s3bucket.S3Bucket"
      }
    }
  }
}
```

Reply:
The problem with logical->physical is that 1 logical can have multiple physical, no?
I'm probably mistaken, so I'll change that

> I also don't think we need the attributes of the resource here.

To prevent attributes from changing to avoid resource replacement. I can remove this if it's not needed.

> terraform_name

I think it's useful for the schema of this pinfile, other than the attributes, to be standard across platforms. Especially since it's not difficult to do. If it's not useful I can remove it.

Reply:
> The problem with logical->physical is that 1 logical can have multiple physical, no?

I think this should refer to the logical path of the *leaf node*.

> I think it's useful for the schema of this pinfile to be standard across platforms

OK, although I am not sure this is something we need to abstract. This name is target specific. For example, in Terraform, it always has the TF type as a prefix, in CFN there are some characters that are not allowed, in K8S it has a strict length limit. In other targets it might not even be a string (god forbid)...

So from a design perspective, it actually makes more sense to give it a concrete name like `terraform_name` or `terraform_id` instead of trying to generalize.

Reply:
> I think this should refer to the logical path of the leaf node.

Yeah that's what I meant. I think platforms can technically emit more than one physical thing given a single logical leaf.

Reply:
Eventually, in CDKTF/AWSCDK/CDK8S there's a 1:1 mapping between L1s and resources in the target configuration. That's the "leaf node" that we should pin.

Reply:
> Eventually, in CDKTF/AWSCDK/CDK8S there's a 1:1 mapping between L1s and resources in the target configuration

I understand that's how these CDKs work, but it is _technically_ possible for a single construct to spit out multiple things. Or I could even make my own type of CDK and not consider that restriction either. I thought it'd be useful for this pinfile to basically skip any possible abstractions and go straight to the actual output both for the source and destination of the mapping.

That being said, I'll change it to logical -> physical because it's obviously the 0.01% case

btw, are you okay with storing the attributes in the pinfile as well or should that be addressed?